docs/proposals/improved-schema-for-services.md

```diff
@@ -0,0 +1,465 @@
+    },
+    {
+      "description": "Limits describes the minimum/maximum amount of compute resources required/allowed",
+      "displayName": "Kafka Resource Requirements",
+      "path": "kafka.resources",
+      "x-descriptors": [
+        "urn:alm:descriptor:com.tectonic.ui:resourceRequirements"
+      ]
+    },
+    {
+      "description": "The desired number of Zookeeper nodes.",
+      "displayName": "Zookeeper Nodes",
+      "path": "zookeeper.replicas",
+      "x-descriptors": [
+        "urn:alm:descriptor:com.tectonic.ui:podCount"
+      ]
+    },
+    {
+      "description": "The type of storage used by Zookeeper nodes",
+      "displayName": "Zookeeper storage",
+      "path": "zookeeper.storage.type",
+      "x-descriptors": [
+        "urn:alm:descriptor:com.tectonic.ui:select:ephemeral",
+        "urn:alm:descriptor:com.tectonic.ui:select:persistent-claim",
+        "urn:alm:descriptor:com.tectonic.ui:select:jbod"
+      ]
+    },
+    {
+      "description": "Limits describes the minimum/maximum amount of compute resources required/allowed",
+      "displayName": "Zookeeper Resource Requirements",
+      "path": "zookeeper.resources",
+      "x-descriptors": [
+        "urn:alm:descriptor:com.tectonic.ui:resourceRequirements"
+      ]
+    }
+  ]
+}
+```
+
+## Allowing user to set the parameters 
+
+The current approach of sending map values via cli are as follows - 
+
+- we would add a `-p` cobra param as a list
+- each of the parameter would represent the key in a map and value in a map `e.g. -p "key"="value"`
+- we would allow json path in the key for the user to specific any field in the map that they want to set e.g. `services[0].namespace`.
+- we optimistically try to convert the type of param values in the below order
+    - try to parse as integer, if it goes through then we assume it as an integer
+    - try to parse as a float and if it goes through the we assume the value as float
+    - try to parse as a boolean and if goes through then we assume the value as boolean
+    - then everything else becomes string.
+
+Sample - `odo service create servicebinding.coreos.io/Servicebinding/<version> -p  "services[0].envVarPrefix"="SVC" -p "services[0].namespace"="openshift"`
+
+We would using https://github.com/tidwall/sjson to map the keys of the json.
+
+this would yield into a map that looks like this
+
+```
+
+{
+  "services":[
+    {
+      "envVarPrefix": "SVC",
+      "namespace": "openshift"
+    }
+  ]
+}
+
+```
+
+- odo also needs to have smart auto-complete which auto selects the version if the CR only has one version.
+
+
+## Using the metadata to validate the user input
+
+At this stage the user either has access to the `openAPIV3Schema` or `ClusterServiceVersion` and also the user has provided the service parameters they want to set as well. To hide the difference between these implementation from the user the `catalog describe` output would follow the same format. e.g.
+```
+- FieldPath: zookeeper.resources
```

Comment:
how are arrays of objects going to be represented in this output?

lets say that there is that `first.second.envVars` is and array of env variable objects (`{name: "foo", value:"bar"}`)
How the `FieldPath` will look like, so users know that `envVars` is an array of objects and the format they need to use is `first.second.envVars[0].name=foo` `first.second.envVars[0].value=bar`)
and they don't try to do something like  `first.second.envVars.name=foo` `first.second.envVars.value=bar` or `first.second[0].envVars.name=foo` `first.second[0].envVars.value=bar`)

Reply:
my thought was to Maybe we can show it as `secret[*].name` and explain in describe that * needs to be an index 0,1….?

Reply:
> my thought was to Maybe we can show it as `first.second.envVars[*].name` and explain in describe that * needs to be an index 0,1….?

:+1: If there is enough information to do that that would be great!

Reply:
yeah we have information on what is an array from swagger

Reply:
added conversion detail in the proposal